fetcher/mitre/mitre.go

```diff
@@ -0,0 +1,675 @@
+package mitre
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/vulsio/go-cve-dictionary/fetcher"
+	"github.com/vulsio/go-cve-dictionary/log"
+	"github.com/vulsio/go-cve-dictionary/models"
+	"github.com/vulsio/go-cve-dictionary/util"
+)
+
+const repositoryURL = "https://github.com/CVEProject/cvelistV5/archive/refs/heads/main.tar.gz"
+
+func Fetch() (string, error) {
+	d := util.CacheDir()
+
+	if err := os.RemoveAll(d); err != nil {
+		return "", xerrors.Errorf("Failed to remove %s. err: %w", d, err)
+	}
+
+	bs, err := fetcher.FetchFeedFile(repositoryURL, false)
+	if err != nil {
+		return "", xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+
+	gr, err := gzip.NewReader(bytes.NewReader(bs))
+	if err != nil {
+		return "", xerrors.Errorf("Failed to create gzip reader. err: %w", err)
+	}
+	defer gr.Close()
+
+	tr := tar.NewReader(gr)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return "", xerrors.Errorf("Failed to next tar reader. err: %w", err)
+		}
+
+		if hdr.FileInfo().IsDir() {
+			continue
+		}
+
+		if !strings.HasPrefix(filepath.Base(hdr.Name), "CVE-") {
+			continue
+		}
+
+		if err := func() error {
+			ss := strings.Split(filepath.Base(hdr.Name), "-")
+			if len(ss) != 3 {
+				return xerrors.Errorf("Failed to parse year. err: invalid ID format. expected: %q, actual: %q", "CVE-yyyy-\\d{4,}.json", filepath.Base(hdr.Name))
+			}
+			if _, err := time.Parse("2006", ss[1]); err != nil {
+				return xerrors.Errorf("Failed to parse year. err: invalid ID format. expected: %q, actual: %q", "CVE-yyyy-\\d{4,}.json", filepath.Base(hdr.Name))
+			}
+
+			if err := os.MkdirAll(filepath.Join(d, ss[1]), os.ModePerm); err != nil {
+				return xerrors.Errorf("Failed to mkdir %s. err: %w", filepath.Join(d, ss[1]), err)
+			}
+
+			f, err := os.Create(filepath.Join(d, ss[1], filepath.Base(hdr.Name)))
+			if err != nil {
+				return xerrors.Errorf("Failed to create %s. err: %w", filepath.Join(d, ss[1], filepath.Base(hdr.Name)), err)
+			}
+			defer f.Close()
+
+			if _, err := io.Copy(f, tr); err != nil {
+				return xerrors.Errorf("Failed to copy %s to %s. err: %w", hdr.Name, filepath.Join(d, ss[1], filepath.Base(hdr.Name)), err)
+			}
+
+			return nil
+		}(); err != nil {
+			return "", xerrors.Errorf("Failed to extract %s. err: %w", hdr.Name, err)
+		}
+	}
+
+	return d, nil
+}
+
+// Convert convert <cacheDir>/<year>/CVE-<year>-\d{4,}.json to []models.Mitre
+func Convert(cacheDir, year string) ([]models.Mitre, error) {
+	var ps []string
+	if err := filepath.WalkDir(cacheDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		if strings.HasPrefix(filepath.Base(path), fmt.Sprintf("CVE-%s-", year)) {
```

Comment:
The prefix string is identical in a function call.
How about store it to a variable before WalkDir() in order to not generate the same string hundreds of thousand times?

Reply:
638ba0c42679e3f28b603e3936c138e8ef577032